### PR TITLE
Adding container-inherit-properties to model-config

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

container-inherit-properties allows a user to specify cloudinit settings which should be replicated from a machine to a containers which juju has provisioned.  

## QA steps

1. Configure MaaS to setup apt settings, mirror or proxy etc. 
2. Bootstrap and provision a machine.
3. juju model-config container-inherit-properties="apt-primary"
4. add a container, lxd and kvm should be used.
5. verify the container picked up the apt cloud init properties

Try 2 series, one of them trusty, host and container with the same series.
Only ca-certs, apt-security, apt-sources and apt-primary allowed.

## Documentation changes

Documentation updates will be needed.

## Bug reference

n/a

### CODE REVIEW

**Comment by hmlanigan** [comment]
[cloudinit phase 2](https://trello.com/c/dTsKm53E/98-cloudinit-phase-2)

File: apiserver/facades/agent/provisioner/provisioner_test.go [hunk 1]
```diff
@@ -1351,11 +1351,12 @@ func (s *withoutControllerSuite) TestContainerManagerConfig(c *gc.C) {
...
-		"apt-https-proxy":       "https://proxy.example.com:9000",
-		"allow-lxd-loop-mounts": true,
-		"apt-mirror":            "http://example.mirror.com",
-		"cloudinit-userdata":    validCloudInitUserData,
+		"http-proxy":                   "http://proxy.example.com:9000",
+		"apt-https-proxy":              "https://proxy.example.com:9000",
+		"allow-lxd-loop-mounts":        true,
+		"apt-mirror":                   "http://example.mirror.com",
+		"cloudinit-userdata":           validCloudInitUserData,
+		"container-inherit-properties": "write_files,apt",
```
Comment (wallyworld) [inline apiserver/facades/agent/provisioner/provisioner_test.go]: At first read, this doesn't see quite right. If the inherited properties say that the container should use the apt mirror from the host for example, then we should be setting apt-mirror here. Same with apt proxy etc. Or if we are to use the CA certs from the host, then there should be a CA certs entry.

File: cloudconfig/containerinit/containerinit_fromhost.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	"sort"
+	"strings"
+
+	"github.com/juju/errors"
+	utilsos "github.com/juju/utils/os"
+	utilsseries "github.com/juju/utils/series"
+	"gopkg.in/yaml.v2"
+
+	"github.com/juju/juju/juju/paths"
+	"github.com/juju/utils"
```
Comment (wallyworld) [inline cloudconfig/containerinit/containerinit_fromhost.go]: move me

File: cloudconfig/containerinit/containerinit_fromhost.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	fileInfo, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot determine files in CloudInitCfgDir for the machine")
+	}
+	sortedFileInfos := sortableFileInfos(fileInfo)
+	sort.Sort(sortedFileInfos)
+	cloudInit := make(map[string]interface{})
+	for _, file := range fileInfo {
+		name := file.Name()
+		if !strings.HasSuffix(name, "cfg") {
```
Comment (wallyworld) [inline cloudconfig/containerinit/containerinit_fromhost.go]: .cfg?

File: cloudconfig/containerinit/containerinit_fromhost.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+		// vendor-data.txt is sometimes empty
+		return nil, nil
+	}
+	// The userdata maybe be gzip'd, base64 encoded, both, or neither.
+	// If both, it's been gzip'd, then base64 encoded.
+	rawBuf, err := unmarshallContainerCloudInit(data)
+	if err == nil {
+		return rawBuf, nil
+	}
+	logger.Debugf("unmarshal of %q failed (%s), maybe it is compressed", file, err)
```
Comment (wallyworld) [inline cloudconfig/containerinit/containerinit_fromhost.go]: i think this is more of a Trace thing

File: cloudconfig/containerinit/containerinit_fromhost.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+	zippedData, err := utils.Gunzip(data)
+	if err == nil {
+		return unmarshallContainerCloudInit(zippedData)
+	}
+	logger.Debugf("Gunzip of %q failed (%s), maybe it is encoded", file, err)
```
Comment (wallyworld) [inline cloudconfig/containerinit/containerinit_fromhost.go]: ditto

File: cloudconfig/containerinit/containerinit_fromhost.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+	decodedData, err := base64.StdEncoding.DecodeString(string(data))
+	if err == nil {
+		// it could still be gzip'd.
+		buf, err := unmarshallContainerCloudInit(decodedData)
+		if err == nil {
+			return buf, nil
+		}
+	}
+	logger.Debugf("Decoding of %q failed (%s), maybe it is encoded and gzipped", file, err)
```
Comment (wallyworld) [inline cloudconfig/containerinit/containerinit_fromhost.go]: ditto

File: cloudconfig/containerinit/containerinit_fromhost.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
+package containerinit
```
Comment (wallyworld) [inline cloudconfig/containerinit/containerinit_fromhost.go]: This would better be named as cloudinit_fromhost.go
The single exported method in this file is to read the cloud init data

File: cloudconfig/containerinit/containerinit_fromhost_test.go [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+import (
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/cloudconfig/containerinit"
+	"github.com/juju/juju/testing"
+	gc "gopkg.in/check.v1"
```
Comment (wallyworld) [inline cloudconfig/containerinit/containerinit_fromhost_test.go]: fix import grouping

File: environs/config/config.go [hunk 9]
```diff
@@ -177,10 +178,16 @@ const (
...
 	FanConfig = "fan-config"
 
-	// CloudInitUserDataKey is the key to specify cloud-init yaml the user wants to add
-	// into the cloud-config data produced by Juju when provisioning machines.
+	// CloudInitUserDataKey is the key to specify cloud-init yaml the user
+	// wants to add into the cloud-config data produced by Juju when
+	// provisioning machines.
 	CloudInitUserDataKey = "cloudinit-userdata"
 
+	// ContainerInheritProperiesKey is the key to specify a list of keys to
```
Comment (wallyworld) [inline environs/config/config.go]: As discussed, let's reword this to remove the cloud init specific implementation aspects

File: environs/config/config.go [hunk 10]
```diff
@@ -634,6 +642,23 @@ func Validate(cfg, old *Config) error {
 		}
 	}
 
+	if raw, ok := cfg.defined[ContainerInheritProperiesKey].(string); ok && raw != "" {
+		rawProperties := strings.Split(raw, ",")
+		propertySet := set.NewStrings()
+		for _, prop := range rawProperties {
+			propertySet.Add(strings.TrimSpace(prop))
+		}
+		if propertySet.Contains("users") {
```
Comment (wallyworld) [inline environs/config/config.go]: I think we need a whitelist of supported attributes. Plus a blacklist of disallowed cloud init passthrough attributes.

File: worker/provisioner/broker.go [hunk 11]
```diff
@@ -11,10 +11,13 @@ import (
 	"gopkg.in/juju/names.v2"
 
 	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/cloudconfig/containerinit"
 	"github.com/juju/juju/container"
 	"github.com/juju/juju/instance"
 	"github.com/juju/juju/network"
 	"github.com/juju/juju/tools"
+	"github.com/juju/utils/set"
+	"strings"
```
Comment (wallyworld) [inline worker/provisioner/broker.go]: import grouping

File: apiserver/facades/agent/provisioner/provisioner_test.go (hunk 1 above)
Comment (hmlanigan) [reply apiserver/facades/agent/provisioner/provisioner_test.go]: updated

File: cloudconfig/containerinit/containerinit_fromhost.go (hunk 7 above)
Comment (hmlanigan) [reply cloudconfig/containerinit/containerinit_fromhost.go]: moved cloudconfig/containerinit/containerinit_fromhost.go to cloudconfig/machinecloudconfig.go

File: cloudconfig/containerinit/containerinit_fromhost.go (hunk 3 above)
Comment (hmlanigan) [reply cloudconfig/containerinit/containerinit_fromhost.go]: done

File: cloudconfig/containerinit/containerinit_fromhost.go (hunk 4 above)
Comment (hmlanigan) [reply cloudconfig/containerinit/containerinit_fromhost.go]: done

File: cloudconfig/containerinit/containerinit_fromhost.go (hunk 5 above)
Comment (hmlanigan) [reply cloudconfig/containerinit/containerinit_fromhost.go]: done

File: cloudconfig/containerinit/containerinit_fromhost.go (hunk 6 above)
Comment (hmlanigan) [reply cloudconfig/containerinit/containerinit_fromhost.go]: done

File: cloudconfig/containerinit/containerinit_fromhost_test.go (hunk 8 above)
Comment (hmlanigan) [reply cloudconfig/containerinit/containerinit_fromhost_test.go]: fixed

File: environs/config/config.go (hunk 9 above)
Comment (hmlanigan) [reply environs/config/config.go]: updated

File: environs/config/config.go (hunk 10 above)
Comment (hmlanigan) [reply environs/config/config.go]: With agreement, a whitelist will be used to allow: "apt-primary", "apt-sources", "apt-security", "ca-certs".  The first 3 values refer to specific second level keys within "apt"

File: worker/provisioner/broker.go (hunk 11 above)
Comment (hmlanigan) [reply worker/provisioner/broker.go]: fixed

**Comment by hmlanigan** [comment]
[cloudinit phase 2](https://trello.com/c/dTsKm53E/98-cloudinit-phase-2)

**Comment by hmlanigan** [comment]
[cloudinit phase 2](https://trello.com/c/dTsKm53E/98-cloudinit-phase-2)

File: cloudconfig/machinecloudconfig.go [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+	"github.com/juju/utils"
+	utilsos "github.com/juju/utils/os"
+	utilsseries "github.com/juju/utils/series"
+	"gopkg.in/yaml.v2"
+
+	"github.com/juju/juju/juju/paths"
+)
+
+// GetMachineCloudInitData returns a map of all cloud init data on the machine.
+func GetMachineCloudInitData(series string) (map[string]interface{}, error) {
```
Comment (wallyworld) [inline cloudconfig/machinecloudconfig.go]: Could we expand the comment to explain how series is used. A reader might wonder why series needs to be specified when it can be determined from the host machine. Maybe a TODO to explain how series will be used when we need to support trusty on xenial etc

File: cloudconfig/machinecloudconfig.go [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+func (fil sortableFileInfos) Swap(i, j int) {
+	fil[i], fil[j] = fil[j], fil[i]
+}
+
+// CloudInitCfgDir is for testing purposes.
+var CloudInitCfgDir = paths.CloudInitCfgDir
+
+// getMachineCloudCfgDirData returns a map of the combined machine's cloud init
```
Comment (wallyworld) [inline cloudconfig/machinecloudconfig.go]: ... machine's combined....

File: cloudconfig/machinecloudconfig.go [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+		}
+	}
+	logger.Tracef("Decoding of %q failed (%s), maybe it is encoded and gzipped", file, err)
+
+	decodedZippedBuf, err := utils.Gunzip(decodedData)
+	if err != nil {
+		// During testing, it was found that the trusty vendor-data.txt.i file
+		// can contain only the text "NONE", which doesn't unmarshall or decompress
+		// we don't want to fail in that case.
+		if series == "trusty" {
```
Comment (wallyworld) [inline cloudconfig/machinecloudconfig.go]: we should check decodedData contents before assuming it's ok to return nil, nil

File: cloudconfig/machinecloudconfig.go [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+				foundDataMap[key] = val
+			} else {
+				log.Debugf("%s not found in machine cloud-init data", key)
+			}
+		}
+	}
+	return foundDataMap
+}
+
+func machineCloudConfigAptV078(key string, val interface{}, log loggo.Logger) map[string]interface{} {
```
Comment (wallyworld) [inline cloudconfig/machinecloudconfig.go]: let's comment this

File: cloudconfig/machinecloudconfig.go [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+			for _, aptKey := range append(aptPrimaryKeys, aptSourcesKeys...) {
+				if val, ok := dataMap[aptKey]; ok {
+					foundDataMap[aptKey] = val
+				} else {
+					log.Debugf("%s not found as part of %s, in machine cloud-init data", strings.Join(keySplit, "-"), key)
+				}
+			}
+		case "apt-security":
+			// Translation for apt-security unknown at this time.
+			log.Debugf("%s not found in machine cloud-init data", key)
```
Comment (wallyworld) [inline cloudconfig/machinecloudconfig.go]: should this be a warning? or return an error? (if the user ha specified something that can't be fulfilled)

File: environs/config/config.go [hunk 17]
```diff
@@ -177,10 +178,16 @@ const (
...
-	// CloudInitUserDataKey is the key to specify cloud-init yaml the user wants to add
-	// into the cloud-config data produced by Juju when provisioning machines.
+	// CloudInitUserDataKey is the key to specify cloud-init yaml the user
+	// wants to add into the cloud-config data produced by Juju when
+	// provisioning machines.
 	CloudInitUserDataKey = "cloudinit-userdata"
 
+	// ContainerInheritProperiesKey is the key to specify a list of properties
+	// to be copied from a machine to a container during provisioning.  The
+	// list will be comma separated.
```
Comment (wallyworld) [inline environs/config/config.go]: Can we add a TODO to use a schema type of List (we are currently working on extending support for config attrs that support list and map types in the schema)

File: environs/config/config.go [hunk 18]
```diff
@@ -634,6 +642,20 @@ func Validate(cfg, old *Config) error {
 		}
 	}
 
+	if raw, ok := cfg.defined[ContainerInheritProperiesKey].(string); ok && raw != "" {
+		rawProperties := strings.Split(raw, ",")
+		propertySet := set.NewStrings()
+		for _, prop := range rawProperties {
+			propertySet.Add(strings.TrimSpace(prop))
+		}
+		whiteListSet := set.NewStrings("apt-primary", "apt-sources", "apt-security", "ca-certs")
```
Comment (wallyworld) [inline environs/config/config.go]: Let's define the supported values as a slice outside this func and use that to init the set

File: worker/provisioner/broker.go [hunk 19]
```diff
@@ -7,10 +7,12 @@ import (
 	"github.com/juju/errors"
 	"github.com/juju/loggo"
 	"github.com/juju/utils/arch"
+	//"github.com/juju/utils/set"
```
Comment (wallyworld) [inline worker/provisioner/broker.go]: remove me

File: cloudconfig/machinecloudconfig.go (hunk 15 above)
Comment (hmlanigan) [reply cloudconfig/machinecloudconfig.go]: done

File: cloudconfig/machinecloudconfig.go (hunk 16 above)
Comment (hmlanigan) [reply cloudconfig/machinecloudconfig.go]: changed to return an error.

File: environs/config/config.go (hunk 17 above)
Comment (hmlanigan) [reply environs/config/config.go]: done

File: environs/config/config.go (hunk 18 above)
Comment (hmlanigan) [reply environs/config/config.go]: done

File: worker/provisioner/broker.go (hunk 19 above)
Comment (hmlanigan) [reply worker/provisioner/broker.go]: done

File: cloudconfig/machinecloudconfig.go (hunk 14 above)
Comment (hmlanigan) [reply cloudconfig/machinecloudconfig.go]: decodedData contents were checked above, lines 143-150, and failed.

File: cloudconfig/machinecloudconfig.go (hunk 13 above)
Comment (hmlanigan) [reply cloudconfig/machinecloudconfig.go]: done

File: cloudconfig/machinecloudconfig.go (hunk 12 above)
Comment (hmlanigan) [reply cloudconfig/machinecloudconfig.go]: done

**Comment by hmlanigan** [comment]
$$merge$$
